src/listener/to_listener.rs

```diff
@@ -0,0 +1,405 @@
+#[cfg(unix)]
+use super::UnixListener;
+use super::{Listener, MultiListener, ParsedListener, TcpListener};
+use crate::http::url::Url;
+use async_std::io;
+use std::net::ToSocketAddrs;
+
+/// ToListener represents any type that can be converted into a
+/// [`Listener`](crate::listener::Listener).  Any type that implements
+/// ToListener can be passed to [`Server::listen`](crate::Server::listen) or
+/// added to a [`MultiListener`](crate::listener::MultiListener)
+///
+/// # Example strings on all platforms include:
+/// * `tcp://localhost:8000`
+/// * `tcp://0.0.0.0` (binds to port 80 by default)
+/// * `http://localhost:8000` (http is an alias for tcp)
+/// * `http://127.0.0.1:8000` (or `0.0.0.0`, or some specific bindable ip)
+/// * `127.0.0.1:3000` (or any string that can be parsed as a [SocketAddr](std::net::SocketAddr))
+/// * `[::1]:1213` (an ipv6 [SocketAddr](std::net::SocketAddr))
+///
+/// # Strings supported only on `cfg(unix)` platforms:
+/// * `unix:///var/run/tide/socket` (absolute path)
+/// * `unix://socket` (relative path)
+/// * `unix://./socket.file` (also relative path)
+/// * `unix://../socket` (relative path)
+/// * any of the above with the alternate schemes of `file://` or `http+unix://`
+///
+/// # String supported only on windows:
+/// * `:3000` (binds to port 3000)
+///
+/// # Specifying multiple listeners:
+/// To bind to any number of listeners concurrently:
+/// ```rust,no_run
+/// # fn main() -> std::io::Result<()> { async_std::task::block_on(async {
+/// # let app = tide::new();
+/// app.listen(vec!["tcp://localhost:8000", "tcp://localhost:8001"]).await?;
+/// # Ok(()) }) }
+/// ```
+///
+/// # Multiple socket resolution
+/// If a TCP listener resolves to multiple socket addresses, tide will
+/// bind to the first successful one. For example, on ipv4+ipv6
+/// systems, `tcp://localhost:1234` resolves both to `127.0.0.1:1234`
+/// (v4) as well as `[::1]:1234` (v6). The order that these are
+/// attempted is platform-determined. To listen on all of the addresses, use
+/// ```rust,no_run
+/// # fn main() -> std::io::Result<()> { async_std::task::block_on(async {
+/// # let app = tide::new();
+/// use std::net::ToSocketAddrs;
+/// app.listen("localhost:8000".to_socket_addrs()?.collect::<Vec<_>>()).await?;
+/// # Ok(()) }) }
+/// ```
+/// # Other implementations
+/// See below for additional provided implementations of ToListener.
+
+pub trait ToListener<State: Send + Sync + 'static> {
+    type Listener: Listener<State>;
+    /// Transform self into a
+    /// [`Listener`](crate::listener::Listener). Unless self is
+    /// already bound/connected to the underlying io, converting to a
+    /// listener does not initiate a connection. An Err return
+    /// indicates an unsuccessful conversion to a listener, not an
+    /// unsuccessful bind attempt.
+    fn to_listener(self) -> io::Result<Self::Listener>;
+}
```

Comment:
One question that I have is: why don't we perform the listening logic inside of the `to_listener` function's body? Intuitively it feels like we may be having one layer of traits more than is required. We can still reuse logic by having an e.g. `listen_unix` function that knows how to listen on a unix socket, but that can be called from this function.

I guess what I'm trying to say is that I'm unclear why "convert into a listener" and "listen on said listener" are two separate steps. My guess is that it's probably coming from the desire to have MultiListener?

During our chat yesterday I showed how to manually `MultiListener`'s functionality using `task::spawn` and `future::join`. We agreed that doing this manually isn't ideal, so having an intermediate abstraction makes sense. But what I think that shows is that servers that have been started *can* in fact be combined after the fact.

I'm not entirely convinced that I've quite covered all the details here -- but I feel like there might be something to it. Would be keen to hear your thoughts on this!

Reply:
I made an experimental branch at https://github.com/jbr/tide/tree/simpler-listener-trait for comparison. Docs haven't been updated, but: There's no longer a ToListener trait, just a [Listener](https://github.com/jbr/tide/blob/simpler-listener-trait/src/listener/to_listener.rs) trait. Calling `Listener::<State>::listen("http://localhost:8000").await` immediately binds, and Listener is implemented for all of the types that ToListener had been implemented for.  MultiListener also binds immediately, though, which feels like a step backwards.  I couldn't figure out how to get it to correctly hold the `Box<dyn Listener>`s. Listen needs to take self because [some listeners](https://github.com/jbr/tide/blob/simpler-listener-trait/src/listener/to_listener.rs#L154-L158) need to move self into the listener.

I still prefer the two-step version, but it's true that there's slightly less code in the one-trait version: https://github.com/jbr/tide/commit/e86e9575f119f2854f58486017364c8b9ffeb619